File: src/tools/featureFlags/update.ts
import { z } from "zod";
import type { Context, Tool } from "../types";
import { UpdateFeatureFlagInputSchema } from "../../schema/flags";
import { getProjectBaseUrl } from "../../lib/utils/api";

const schema = z.object({
	flagKey: z.string(),
	data: UpdateFeatureFlagInputSchema,
});

type Params = z.infer<typeof schema>;

export const updateHandler = async (context: Context, params: Params) => {
	const { flagKey, data } = params;
	const projectId = await context.getProjectId();

	const flagResult = await context.api.featureFlags({ projectId }).update({
		key: flagKey,
		data: data,
	});

	if (!flagResult.success) {
		throw new Error(`Failed to update feature flag: ${flagResult.error.message}`);
	}

	const featureFlagWithUrl = {
		...flagResult.data,
		url: `${getProjectBaseUrl(projectId)}/feature_flags/${flagResult.data.id}`,
	};

	return {
		content: [{ type: "text", text: JSON.stringify(featureFlagWithUrl) }],
	};
};

const tool = (): Tool<typeof schema> => ({
	name: "update-feature-flag",
	description: `Update a new feature flag in the project.
    - To enable a feature flag, you should make sure it is active and the rollout percentage is set to 100 for the group you want to target.
    - To disable a feature flag, you should make sure it is inactive, you can keep the rollout percentage as it is.
    `,
	schema,
	handler: updateHandler,
});

export default tool;

File: python/tools/feature_flags/update.py
import json

from api.client import is_error, is_success
from lib.utils.api import get_project_base_url
from schema.tool_inputs import FeatureFlagUpdateSchema
from tools.types import Context, Tool, ToolResult


async def update_feature_flag_handler(context: Context, params: FeatureFlagUpdateSchema) -> ToolResult:
    project_id = await context.get_project_id()

    key = params.flagKey
    update_data = params.data

    flag_result = await context.api.feature_flags(project_id).update(key, update_data)

    if is_error(flag_result):
        raise Exception(f"Failed to update feature flag: {flag_result.error}")

    assert is_success(flag_result)

    flag_data = flag_result.data

    feature_flag_with_url = {
        **flag_data.model_dump(),
        "url": f"{get_project_base_url(project_id, context.api.base_url)}/feature_flags/{flag_data.id}",
    }

    return ToolResult(content=json.dumps(feature_flag_with_url))


def update_feature_flag_tool() -> Tool[FeatureFlagUpdateSchema]:
    return Tool(
        name="update-feature-flag",
        description="Updates an existing feature flag in the project.",
        schema=FeatureFlagUpdateSchema,
        handler=update_feature_flag_handler,
    )
